### PR TITLE
TMA-483 request number optimisation in synchronize users action

### PR DESCRIPTION
**Pull request by Hahihula**

### CODE REVIEW

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 1]
```diff
@@ -197,14 +197,17 @@ def call(params)
               fail 'The filter set can not be empty when using sync_multiple_projects_* mode as the filters contain \
                     the project ids in which the permissions should be changed'
             end
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            all_clients = domain.clients(:all, data_product).to_a
+            filters.group_by { |u| u[multiple_projects_column] }.map do |client_id, new_filters|
               fail "Client id cannot be empty" if client_id.blank?
-              project = domain.clients(client_id, data_product).project
+              c = all_clients.detect { |specific_client| specific_client.id == client_id }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'all_clients.detect { |specific_client| specific_client.id == client_id }' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'all_clients.detect' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'specific_client.id == client_id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'specific_client.id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 2]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
+              fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\"" if c.nil?
+              project = c.project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'c.project' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -197,14 +197,17 @@ def call(params)
               fail 'The filter set can not be empty when using sync_multiple_projects_* mode as the filters contain \
                     the project ids in which the permissions should be changed'
             end
-            filters.group_by { |u| u[multiple_projects_column] }.flat_map do |client_id, new_filters|
+            all_clients = domain.clients(:all, data_product).to_a
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(:all, data_product)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 4]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
+            filters.group_by { |u| u[multiple_projects_column] }.map do |client_id, new_filters|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by { |u| u[multiple_projects_column] }' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by { |u| u[multiple_projects_column] }.map' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'u[multiple_projects_column]' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'u' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 5]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
               fail "Client id cannot be empty" if client_id.blank?
-              project = domain.clients(client_id, data_product).project
+              c = all_clients.detect { |specific_client| specific_client.id == client_id }
+              fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\"" if c.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 6]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
                         fail "Client id cannot be empty" if client_id.blank?
-                        begin
-                          project = domain.clients(client_id, data_product).project
-                        rescue RestClient::BadRequest => e
-                          raise e unless /does not exist in data product/ =~ e.response
-                          fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\""
-                        end
+                        c = all_clients.detect { |specific_client| specific_client.id == client_id }
+                        fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\"" if c.nil?
+                        project = c.project
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'c.project' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 7]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
-                                           domain: domain,
-                                           whitelists: whitelists,
-                                           ignore_failures: ignore_failures,
-                                           remove_users_from_project: remove_users_from_project,
-                                           do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                           create_non_existing_user_groups: create_non_existing_user_groups)
+                      project.import_users(filtered_users, common_params)
                     when 'sync_multiple_projects_based_on_custom_id'
-                      new_users.group_by { |u| u[:pid] }.flat_map do |client_id, users|
+                      all_clients = domain.clients(:all, data_product).to_a
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'domain.clients(:all, data_product)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 8]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
+                      new_users.group_by { |u| u[:pid] }.map do |client_id, nusers|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'new_users.group_by { |u| u[:pid] }' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'new_users.group_by' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'u[:pid]' 6 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'u' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 9]
```diff
@@ -219,13 +208,7 @@ def call(params)
...
                     when 'sync_one_project_based_on_pid'
                       filtered_users = new_users.select { |u| u[:pid] == project.pid }
-                      project.import_users(filtered_users,
-                                           domain: domain,
-                                           whitelists: whitelists,
-                                           ignore_failures: ignore_failures,
-                                           remove_users_from_project: remove_users_from_project,
-                                           do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                           create_non_existing_user_groups: create_non_existing_user_groups)
+                      project.import_users(filtered_users, common_params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'project.import_users(filtered_users, common_params)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 10]
```diff
@@ -190,25 +191,13 @@ def call(params)
...
                       users.map(&:delete)
                     when 'sync_project'
-                      project.import_users(new_users,
-                                           domain: domain,
-                                           whitelists: whitelists,
-                                           ignore_failures: ignore_failures,
-                                           remove_users_from_project: remove_users_from_project,
-                                           do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                           create_non_existing_user_groups: create_non_existing_user_groups)
+                      project.import_users(new_users, common_params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'project.import_users(new_users, common_params)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 11]
```diff
@@ -190,25 +191,13 @@ def call(params)
...
                           project = client.projects(project_id)
                           fail "You (user executing the script - #{client.user.login}) is not admin in project \"#{project_id}\"." unless project.am_i_admin?
-                          project.import_users(users,
-                                               domain: domain,
-                                               whitelists: whitelists,
-                                               ignore_failures: ignore_failures,
-                                               remove_users_from_project: remove_users_from_project,
-                                               do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                               create_non_existing_user_groups: create_non_existing_user_groups)
+                          project.import_users(users, common_params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'project.import_users(users, common_params)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 12]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
-                        puts "Project #{project.pid} of client #{client_id} will receive #{users.count} users"
-                        project.import_users(users,
-                                             domain: domain,
-                                             whitelists: whitelists,
-                                             ignore_failures: ignore_failures,
-                                             remove_users_from_project: remove_users_from_project,
-                                             do_not_touch_users_that_are_not_mentioned: do_not_touch_users_that_are_not_mentioned,
-                                             create_non_existing_user_groups: create_non_existing_user_groups)
-                      end
+                        puts "Project #{project.pid} of client #{client_id} will receive #{nusers.count} users"
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'project.pid' 7 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 13]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
+                      all_clients = domain.clients(:all, data_product).to_a
+                      new_users.group_by { |u| u[:pid] }.map do |client_id, nusers|
                         fail "Client id cannot be empty" if client_id.blank?
-                        begin
-                          project = domain.clients(client_id, data_product).project
-                        rescue RestClient::BadRequest => e
-                          raise e unless /does not exist in data product/ =~ e.response
-                          fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\""
-                        end
+                        c = all_clients.detect { |specific_client| specific_client.id == client_id }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'specific_client.id == client_id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Calls 'specific_client.id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_users.rb [hunk 14]
```diff
@@ -245,32 +228,18 @@ def call(params)
...
+                        fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\"" if c.nil?
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_users.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/models/project.rb [hunk 15]
```diff
@@ -1601,7 +1602,7 @@ def import_users(new_users, options = {})
       whitelisted_new_users, whitelisted_users = whitelist_users(new_users.map(&:to_hash), users_list, options[:whitelists])
 
       # First check that if groups are provided we have them set up
-      check_groups(new_users.map(&:to_hash).flat_map { |u| u[:user_group] || [] }.uniq, options)
+      @user_groups_cache = check_groups(new_users.map(&:to_hash).flat_map { |u| u[:user_group] || [] }.uniq, @user_groups_cache, options)
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'new_users.map(&:to_hash)' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/project.rb]: Calls 'new_users.map(&:to_hash).flat_map' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/project.rb]: Has the variable name 'u' (UncommunicativeVariableName)

File: lib/gooddata/models/project.rb [hunk 16]
```diff
@@ -1739,8 +1740,9 @@ def remove_users(list)
       end
     end
 
-    def check_groups(specified_groups, options = {})
-      groups = user_groups.map(&:name)
+    def check_groups(specified_groups, user_groups_cache, options = {})
+      user_groups_cache = user_groups if user_groups_cache.nil? || user_groups_cache.empty?
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Performs a nil-check (NilCheck)

File: lib/gooddata/models/project.rb [hunk 17]
```diff
@@ -1739,8 +1740,9 @@ def remove_users(list)
       end
     end
 
-    def check_groups(specified_groups, options = {})
-      groups = user_groups.map(&:name)
+    def check_groups(specified_groups, user_groups_cache, options = {})
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Has approx 7 statements (TooManyStatements)

File: lib/gooddata/models/project.rb [hunk 18]
```diff
@@ -1701,7 +1702,7 @@ def import_users(new_users, options = {})
           user_groups(g).set_members(remote_users)
         end
         mentioned_groups = mappings.map(&:last).uniq
-        groups_to_cleanup = user_groups.reject { |g| mentioned_groups.include?(g.name) }
+        groups_to_cleanup = @user_groups_cache.reject { |g| mentioned_groups.include?(g.name) }
```
Comment (yenkins) [inline lib/gooddata/models/project.rb]: Has the variable name 'g' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/193/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1029/ : SUCCESS in 1m 17s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1298/ : SUCCESS in 52s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/194/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1417/ : FAILURE in 33m 15s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-lcm-integration-tests-zuul-el7/2/ : FAILURE in 33m 27s

**Comment by Hahihula** [comment]
extended test

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/205/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-integration-tests-zuul/1426/ : FAILURE in 31m 11s
- https://checklist.intgdc.com/job/lcm/job/gooddata-ruby-lcm-integration-tests-zuul-el7/11/ : UNSTABLE in 33m 27s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 19]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
+              project = c.project
               fail "Client #{client_id} does not have project." unless project
               filters_to_load = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
               puts "Synchronizing #{filters_to_load.count} filters in project #{project.pid} of client #{client_id}"
               project.add_data_permissions(filters_to_load, run_params)
-            end
+            end.flatten
           when 'sync_domain_client_workspaces'
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'all_clients.detect { |specific_client| specific_client.id == client_id }' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'all_clients.detect' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'specific_client.id == client_id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'specific_client.id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'c' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 20]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
             without_check(PARAMS, params) do
               CSV.foreach(File.open(data_source.realize(params), 'r:UTF-8'), headers: csv_with_headers, return_headers: false, encoding: 'utf-8') do |row|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'c.project' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 21]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
+            all_clients = domain.clients(:all, data_product).to_a
+            filters.group_by { |u| u[multiple_projects_column] }.map do |client_id, new_filters|
               fail "Client id cannot be empty" if client_id.blank?
-              project = domain.clients(client_id, data_product).project
+              c = all_clients.detect { |specific_client| specific_client.id == client_id }
+              fail "The client \"#{client_id}\" does not exist in data product \"#{data_product.data_product_id}\"" if c.nil?
+              project = c.project
               fail "Client #{client_id} does not have project." unless project
               filters_to_load = GoodData::UserFilterBuilder.get_filters(new_filters, symbolized_config)
               puts "Synchronizing #{filters_to_load.count} filters in project #{project.pid} of client #{client_id}"
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'domain.clients(:all, data_product)' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 22]
```diff
@@ -197,14 +197,17 @@ def call(params)
...
               project.add_data_permissions(filters_to_load, run_params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by { |u| u[multiple_projects_column] }' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by { |u| u[multiple_projects_column] }.map' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'filters.group_by' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Calls 'u[multiple_projects_column]' 3 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Has the variable name 'u' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ruby-opensource-repository-check/238/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/1081/ : FAILURE in 1m 33s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/1350/ : FAILURE in 2m 31s

**Pull request closed without merging**